stolon/src/lib.rs

```diff
@@ -0,0 +1,89 @@
+use std::{net::SocketAddr, sync::Arc};
+
+use tokio::net::TcpStream;
+use tokio_socks::tcp::Socks5Stream;
+use tower::{Service, ServiceExt};
+
+pub use zebra_chain::{
+    parameters::Network, serialization::ZcashDeserialize, transaction::Transaction,
+};
+
+pub type BoxError = Box<dyn std::error::Error + Send + Sync + 'static>;
+
+// Tor's IsolateSOCKSAuth option (on by default) prevents sharing circuits
+// for streams for which different SOCKS authentication was provided.
+
+const SOCKS_USERNAME: &str = "stolon";
+fn socks_password() -> String {
+    use rand::distributions::{Alphanumeric, Distribution};
+    Alphanumeric
+        .sample_iter(rand::thread_rng())
+        .take(16)
+        .collect()
+}
+
+/// Send a transaction over Tor to the Zcash network, using DNS seeders to find peers.
+#[tracing::instrument(skip(transaction))]
+pub async fn send_transaction(
+    tor_addr: SocketAddr,
+    network: Network,
+    transaction: Arc<Transaction>,
+) -> Result<(), BoxError> {
+    // We want to ensure that we get a different peer from the DNS seeder for
+    // each call, so that fanout will send multiple times to different peers
+    // rather than multiple times to the same peer. However, Tor only sends a
+    // single IP address per RESOLVED cell, and caches DNS lookups. As a
+    // workaround, instead of looking up an IP and then connecting, try
+    // connecting directly to the DNS seeder, so that DNS resolution is
+    // performed by the exit. Since we use a new circuit for each call, we hit a
+    // different exit each time.
```

Comment:
I just did a quick check to make sure this is true:
* [script](https://gist.github.com/teor2345/dbd45dd1bd372f06702106d8682650df)
* [list](https://gist.github.com/teor2345/a11fcd2e0a8ebe0fc0229cdd8a7517ea)

There are over 600 tor exits that support port 8233 - about half the number of exits:
https://metrics.torproject.org/relayflags.html

So there are enough exits to support this design. (And there are ~300 that support the testnet port, which is enough for testing.)

The largest relay has a weight that's around 1.5% of the total. So with a fanout of 4, the probability of choosing the same exit is at most 5%. (I derived an upper bound for the probability, using the weight of the largest exit. The exact probability is the sum of `p * (1 - p)^3` for all exits, which is bounded above by `(1 - p)^3` for the largest exit.)

I'm not sure if choosing the same Zcash node twice matters for privacy here. But if it does, we should adjust the fanout to compensate.

Reply:
I don't think that choosing the same Zcash node twice matters for privacy, because that node doesn't learn any new information about the client or the transaction.  The motivation for doing fanout is just to reduce the probability of failure in transaction diffusion.

Reply:
Ok, so it is very unlikely that we'll choose the same node 3 times, and that node is also failing.

It's also unlikely that we'll choose multiple nodes, and they are all failing.